src/main/java/io/anserini/index/IndexUtils.java

```diff
@@ -342,37 +348,69 @@ public void dumpRawDocuments(String reqDocidsPath, boolean prependDocid) throws
 
     InputStream in = getReadFileStream(reqDocidsPath);
     BufferedReader bRdr = new BufferedReader(new InputStreamReader(in));
-    FileOutputStream fOut = new FileOutputStream(new File(reqDocidsPath+".output.tar.gz"));
-    BufferedOutputStream bOut = new BufferedOutputStream(fOut);
-    GzipCompressorOutputStream gzOut = new GzipCompressorOutputStream(bOut);
-    TarArchiveOutputStream tOut = new TarArchiveOutputStream(gzOut);
-
-    String docid;
-    int counter = 0;
-    while ((docid = bRdr.readLine()) != null) {
-      counter += 1;
-      Document d = reader.document(convertDocidToLuceneDocid(docid));
-      IndexableField doc = d.getField(LuceneDocumentGenerator.FIELD_RAW);
-      if (doc == null) {
-        throw new NotStoredException("Raw documents not stored!");
+    
+    String outputDir = Paths.get(reqDocidsPath).getFileName()+"_rawdocs.dump";
+    boolean mkdirStatus = new File(outputDir).mkdirs();
+    if (!mkdirStatus) {
+      LOG.info("Create dump directory failed: "+ outputDir);
+      return;
+    }
+    final class DumpThread extends Thread {
+      final private IndexReader reader;
+      final private String docid;
+      final private String outputDir;
+      final private boolean prependDocid;
+      public DumpThread(IndexReader reader, String docid, String outputDir, boolean prependDocid) throws IOException {
+        this.reader = reader;
+        this.docid = docid;
+        this.outputDir = outputDir;
+        this.prependDocid = prependDocid;
       }
-      TarArchiveEntry tarEntry = new TarArchiveEntry(new File(docid));
-
-      byte[] bytesOut = doc.stringValue().getBytes(StandardCharsets.UTF_8);
-      tarEntry.setSize(bytesOut.length + (prependDocid ? String.format("<DOCNO>%s</DOCNO>\n", docid).length() : 0));
-      tOut.putArchiveEntry(tarEntry);
-      if (prependDocid) {
-        tOut.write(String.format("<DOCNO>%s</DOCNO>\n", docid).getBytes());
+      @Override
+      public void run() {
+        try {
+          Document d = reader.document(convertDocidToLuceneDocid(docid));
+          IndexableField doc = d.getField(LuceneDocumentGenerator.FIELD_RAW);
+          if (doc == null) {
+            LOG.error("Raw documents not stored: " + docid);
```

Comment:
This should be an exception that bombs out of the util, no?
Without the docs stored you really can't do anything...